File: backend/src/main/java/it/chalmers/gamma/filter/ApiKeyAuthenticationFilter.java
package it.chalmers.gamma.filter;

import it.chalmers.gamma.apikey.ApiKeyService;
import it.chalmers.gamma.user.service.UserService;

import java.io.IOException;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.filter.OncePerRequestFilter;



public class ApiKeyAuthenticationFilter extends OncePerRequestFilter {

    private final ApiKeyService apiKeyService;
    private final UserService userService;

    public ApiKeyAuthenticationFilter(ApiKeyService apiKeyService, UserService userService) {
        this.apiKeyService = apiKeyService;
        this.userService = userService;
    }

    /*
    Authentication using API keys are not supported natively, so once a recognized API key is sent, the Server
    authenticates the session by giving the request the Principal and Authorities of the Admin user.
     */
    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain) throws ServletException, IOException {
        String token = resolveToken(request);
        if (this.apiKeyService.isValidApiKey(token)) {
            Authentication auth = getAdminAuthentication();
            SecurityContextHolder.getContext().setAuthentication(auth);
        }
        filterChain.doFilter(request, response);
    }

    private String resolveToken(HttpServletRequest req) {
        String basicToken = req.getHeader("Authorization");
        if (basicToken != null && basicToken.startsWith("pre-shared ")) {
            return removeBasic(basicToken);
        }
        return null;
    }
    // THIS IS SOOOOOOO UGLY
    // Can't figure out another, less hacky way to do it...
    private Authentication getAdminAuthentication() {
        UserDetails userDetails = this.userService.loadUserByUsername("admin");
        return new UsernamePasswordAuthenticationToken(userDetails.getUsername(), null, userDetails.getAuthorities());
    }

    private String removeBasic(String token) {
        return token.substring(11);
    }
}

File: app/src/main/kotlin/it/chalmers/gamma/ApiKeyAuthenticationFilter.kt
package it.chalmers.gamma

import it.chalmers.gamma.apiaccess.ApiCredentialAuthenticator
import it.chalmers.gamma.apiaccess.ApiKey
import it.chalmers.gamma.apiaccess.ApiKeyId
import it.chalmers.gamma.apiaccess.RawApiToken
import it.chalmers.gamma.platform.core.Actor
import it.chalmers.gamma.platform.core.ApiClientCredentialId
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.filter.OncePerRequestFilter

internal class ApiKeyAuthenticationFilter(
    private val authenticator: ApiCredentialAuthenticator,
) : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val credentials = request.getHeader("Authorization")?.toApiKeyCredentials()
        val key = credentials?.let { authenticator.authenticate(it.id, it.token) }
        if (key != null) {
            val principal =
                AuthenticatedApiKey(
                    key,
                    Actor.ApiClient(ApiClientCredentialId(key.id.value)),
                )
            SecurityContextHolder.getContext().authentication =
                UsernamePasswordAuthenticationToken.authenticated(principal, null, emptyList())
        }
        filterChain.doFilter(request, response)
    }
}

internal data class AuthenticatedApiKey(
    val key: ApiKey,
    val actor: Actor.ApiClient,
) {
    override fun toString(): String = "AuthenticatedApiKey(<redacted>)"
}

private data class ApiKeyCredentials(
    val id: ApiKeyId,
    val token: RawApiToken,
)

private fun String.toApiKeyCredentials(): ApiKeyCredentials? {
    if (!startsWith("pre-shared ")) return null
    val credential = removePrefix("pre-shared ")
    val separator = credential.indexOf(':')
    if (separator < 1) return null
    val id = runCatching { ApiKeyId.parse(credential.substring(0, separator)) }.getOrNull() ?: return null
    val token = runCatching { RawApiToken(credential.substring(separator + 1)) }.getOrNull() ?: return null
    return ApiKeyCredentials(id, token)
}
